core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1464,6 +1464,21 @@ class UnifiedLog(@volatile var logStartOffset: Long,
     }
   }
 
+  /**
+   * Check if this partition has, or will have remote logs, so that we should not delete local logs immediately.
+   * We will make sure `remoteLogEnabled` and:
+   *  (1) remote copy is enabled, which means local logs will keep uploading to the remote storage.
+   *  (2) remote copy is disabled, which means there will be no more logs uploaded to the remote storage.
+   *      And check (log start offset != local log start offset), which means the remote storage is not empty. In this case,
+   *      we delete local logs that are already uploaded to the remote storage. And then, when remote storage is empty
+   *      (log start offset == local log start offset) we can treat this topic as "remote log is disabled" and start to delete local logs.
+   *
+   * @return true if this partition has (or will have) remote logs
+   */
+  private def hasRemoteLogs(): Boolean = {
+    remoteLogEnabled() && (!config.remoteCopyDisabled() || logStartOffset != localLogStartOffset())
```

Comment:
`logStartOffset` may not converge with `localLogStartOffset` if there is a leader change post disabling the remote copy since the segment roll-over are independent of the replicas.

Reply:
hmm... any better idea to check it?

Reply:
`logStartOffset` >= `localLogStartOffset` ?

Reply:
`logStartOffset` >= `highestOffsetInRemoteStorage`?

Reply:
Sorry for the confusion. The existing check itself seems sufficient since we always move the LLSO to be the max (LSO, LLSO).

https://sourcegraph.com/github.com/apache/kafka@trunk/-/blob/core/src/main/scala/kafka/log/UnifiedLog.scala?L173

Reply:
Awesome! Thank you!